pkg/report/custom_app_metrics_report.go

```diff
@@ -0,0 +1,68 @@
+package report
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+var customAppMetricsReportMtx = sync.Mutex{}
+
+type CustomAppMetricsReport struct {
+	Events []map[string]interface{} `json:"events"`
+}
+
+func (r *CustomAppMetricsReport) GetType() ReportType {
+	return ReportTypeCustomAppMetrics
+}
+
+func (r *CustomAppMetricsReport) GetSecretName() string {
+	return fmt.Sprintf(ReportSecretNameFormat, r.GetType())
+}
+
+func (r *CustomAppMetricsReport) GetSecretKey() string {
+	return ReportSecretKey
+}
+
+func (r *CustomAppMetricsReport) AppendEvents(report Report) error {
+	reportToAppend, ok := report.(*CustomAppMetricsReport)
+	if !ok {
+		return errors.Errorf("report is not a custom app metrics report")
+	}
+
+	r.Events = append(r.Events, reportToAppend.Events...)
+	if len(r.Events) > r.GetEventLimit() {
+		r.Events = r.Events[len(r.Events)-r.GetEventLimit():]
+	}
+
+	// remove one event at a time until the report is under the size limit
+	encoded, err := EncodeReport(r)
+	if err != nil {
+		return errors.Wrap(err, "failed to encode report")
+	}
+	for len(encoded) > r.GetSizeLimit() {
```

Comment:
is it worth also doing this for the other report types? while we likely are ok, it could be a safeguard against instances with potentially large data in `resource_states`?

Reply:
hmm, yeah. i will update.